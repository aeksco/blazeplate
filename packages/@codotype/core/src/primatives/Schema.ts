import { v4 as uuidv4 } from "uuid";
import { SchemaInput, Attribute, RelationInput } from "..";
import { CreatedBy, CreatedByValues } from "../created-by";
import { TokenPluralization } from "../token";
import { ConfigurationValue } from "../configuration-value";

// // // //

interface SchemaBuilderParams {
    id?: string;
    attributes: Attribute[];
    relations: RelationInput[];
    identifiers: TokenPluralization;
    source?: CreatedBy;
    locked?: boolean;
    configuration?: ConfigurationValue;
    internalNote?: string;
}

export class SchemaBuilder implements SchemaInput {
    id: string = uuidv4();
    attributes: Attribute[];
    relations: RelationInput[];
    identifiers: TokenPluralization;
    createdBy: CreatedBy = CreatedByValues.user;
    locked: boolean = false;
    configuration: ConfigurationValue = {};
    internalNote: string = "";

    constructor(params: SchemaBuilderParams) {
        this.attributes = params.attributes;
        this.relations = params.relations;
        this.identifiers = params.identifiers;

        this.id = params.id || this.id;
        this.createdBy = params.source || this.createdBy;
        this.locked = params.locked !== undefined ? params.locked : this.locked;
        this.configuration = params.configuration || this.configuration;
        this.internalNote = params.internalNote || this.internalNote;
    }
}
