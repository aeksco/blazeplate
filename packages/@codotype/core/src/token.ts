/**
 * TokenPluralization
 * Encapsulates both singular and plural TokenCasing objects - @see Schema.identifiers
 * @param singular - the singular TokenCasing object
 * @param plural - the plural TokenCasing object
 */
export interface TokenPluralization {
    singular: TokenCasing;
    plural: TokenCasing;
}

/**
 * TokenCasing
 * Encapsulates different tokens used by a Plugin to produce code that matches situational expectations
 * @param label - the title-cased token, i.e. `My Project`
 * @param snake - the snake-cased token, i.e. `my_project`
 * @param camel - the camel-cased token, i.e. `myProject`
 * @param pascal - the pascal-cased token, i.e. `MyProject`
 * @param kebab - the kebab-cased token, i.e. `my-project`
 */
export interface TokenCasing {
    label: string;
    snake: string;
    camel: string;
    pascal: string;
    kebab: string;
}

/**
 * EMPTY_TOKEN_CASING
 * Defines an empty TokenCasing object
 */
export const EMPTY_TOKEN_CASING: TokenCasing = {
    label: "",
    snake: "",
    camel: "",
    pascal: "",
    kebab: "",
};
