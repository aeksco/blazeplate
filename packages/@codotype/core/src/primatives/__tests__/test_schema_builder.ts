import { SchemaBuilder } from "../Schema";
import { EMPTY_TOKEN_CASING } from "../../token";

// Mocks UUID module
jest.mock("uuid", () => ({
    __esModule: true,
    v4: () => "1234-5678-1234-5678",
}));

// // // //

describe("factory", () => {
    test("works", () => {
        const schema = new SchemaBuilder({
            attributes: [],
            relations: [],
            identifiers: {
                singular: {
                    ...EMPTY_TOKEN_CASING,
                },
                plural: {
                    ...EMPTY_TOKEN_CASING,
                },
            },
        });

        expect(schema).toMatchSnapshot();
    });
});
