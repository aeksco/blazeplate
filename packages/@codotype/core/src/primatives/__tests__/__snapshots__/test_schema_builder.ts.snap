// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`factory works 1`] = `
SchemaBuilder {
  "attributes": Array [],
  "configuration": Object {},
  "id": "1234-5678-1234-5678",
  "identifiers": Object {
    "plural": Object {
      "camel": "",
      "kebab": "",
      "pascal": "",
      "snake": "",
      "title": "",
    },
    "singular": Object {
      "camel": "",
      "kebab": "",
      "pascal": "",
      "snake": "",
      "title": "",
    },
  },
  "internalNote": "",
  "locked": false,
  "relations": Array [],
  "removable": true,
  "source": "user",
}
`;
